### PR TITLE
[AIRFLOW-276] Refresh stale dags

### PR DESCRIPTION
**Pull request by ldct**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-276

Parsing and executing dag files can be slow, since they are python scripts. Hence, we cannot reload the dags folder on every request. The current code deals with this by only loading the dags folder once on startup, so it doesn't pick up new changes.

This PR starts, for each flask (request-serving) worker process, a thread that continuously loads the dags folder. Even if that is slow, the main thread will still serve requests quickly. The threads share a single dagbag object, and once the child thread finishes loading the dags folder, it updates the dagbag object.

There are a few alternative strategies:

Option 1. We could call `refresh_dags` on every page view. Most of the time the file timestamps will be unchanged and this will be fast. However if an expensive-to-load dag is added, an unlucky user will have to wait a few seconds for his request to be served.
Option 2. We could make the webserver code not load the dags folder at all, and read exclusively from the airflow db. 

Concretely, this option means
1. A new dag is uploaded to the scheduler's dag folder, or an existing dag is updated
2. The scheduler loads the new dag
3. The scheduler partially serialized the new dag and places it in the DB (it cannot be fully serialized since dags are not in general pickleable)
4. The web UI only renders from the db (it does not need to have the .py file), and so picks up the change

Disadvantages of option 2:
1. Dags must know how to partially serialize themselves in sufficient detail that the web UI can display all the existing details about them
2. More data must be put into the db than is currently there
3. If the scheduler is not up, the UI will not be updated.

Option 3. A separate process (instead of thread) could be started on the web UI, which passes a partially serialized dag to the web process. This also has the disadvantage that dags must know how to partially serialize themselves.
### Testing Done:
- [ ] start webserver, add a new dag to `~/airflow/dags`, make sure the new change is visible in the web UI
- [ ] add a new dag with a syntax error to `~/airflow/dags`, make sure the web UI is still up

@plypaul @mistercrunch

### CODE REVIEW

File: airflow/utils/timeout.py [hunk 1]
```diff
@@ -39,8 +40,9 @@ def handle_timeout(self, signum, frame):
 
     def __enter__(self):
         try:
-            signal.signal(signal.SIGALRM, self.handle_timeout)
-            signal.alarm(self.seconds)
+            if isinstance(threading.current_thread(), threading._MainThread):
```
Comment (plypaul) [inline airflow/utils/timeout.py]: If a caller tries to use this outside of the main thread, should we throw an exception?

File: airflow/models.py [hunk 2]
```diff
@@ -180,6 +178,14 @@ def size(self):
         """
         return len(self.dags)
 
+    def refresh_dags(self):
+        if self.include_examples:
+            example_dag_folder = os.path.join(
+                os.path.dirname(__file__),
+                'example_dags')
+            self.collect_dags(example_dag_folder)
```
Comment (plypaul) [inline airflow/models.py:186]: Is `collect_dags` thread safe?

File: airflow/utils/timeout.py (hunk 1 above)
Comment (ldct) [inline airflow/utils/timeout.py]: We'd have to make `process_file` stop using timeout if it's not on the main thread (https://github.com/apache/incubator-airflow/blob/master/airflow/models.py#L252)

**Comment by bolkedebruin** [comment]
Hey @zodiac . Thanks for the contrbution, but I do think you are overlooking another approach. See https://github.com/apache/incubator-airflow/pull/1603 (unfinished at the moment), which is already addressing this issue. It is your no 2, but without the downside. Furthermore, with threading (which is not real multi-threading) you get issues with "timeout" that only works on the main thread.

**Comment by ldct** [comment]
@bolkedebruin I took a look at #1603, it's not really the same as my option 2. AFAICT #1603 loads out of date dags on the webserver thread, which is similar to option 1 (since refresh_dags checks timestamps to only parse changed dags). It had the same drawback that if a slow-to-parse dag is added, some user request will be blocked on loading the dag.

**Comment by bolkedebruin** [comment]
As you are using threading (which is cooperative) and not multiprocessing the issue will remain the same for dags that are difficult to parse or am I incorrect?

The approach in #1603 indeed uses the the DB to determine if the dags need to be refreshed. This is a combination of 1 and 2 without the downside of 1 polling the filesystem for every request. Also if one of the webserver workers refreshes it will propagate to all other webserver workers, thus #1603 deals with the issue of outdated DAGs in the view (mismatch between the scheduler and webserver). 

I dont particularly like the constant polling approach that you are suggesting. This increases load on the system (atimes being updated on the fs, stats etc). In addition in increases the chances of a state mismatch between the scheduler and webserver (now suddenly the webserver can have newer definitions than the scheduler, who has what?) I do understand the "parsing of complex dags" issue, but I wonder if it is the right way of doing it. 

I would support a combination between #1603 and this. That would solve two issues from the user's perspective and reduce redundant polling and keep possible mismatches between the scheduler and webserver as low as possible.

**Comment by ldct** [comment]
@bolkedebruin 

> As you are using threading (which is cooperative) and not multiprocessing the issue will remain the same for dags that are difficult to parse or am I incorrect?

No, threading is preemptive, I tested to confirm by adding a dag that never finishes parsing into the dag folder while the webserver is up. The webserver continues to respond to handle requests and only the background thread is blocked on parsing.

**Comment by ldct** [comment]
@bolkedebruin 

> In addition in increases the chances of a state mismatch between the scheduler and webserver (now suddenly the webserver can have newer definitions than the scheduler, who has what?)

This can already happen now, if gunicorn decides to restart the webserver when the webserver has newer definitions than the scheduler.

File: airflow/utils/timeout.py (hunk 1 above)
Comment (plypaul) [inline airflow/utils/timeout.py]: Well as it is now, this passes silently, right?

Comment (ldct) [inline airflow/utils/timeout.py]: Yeah, no signal handler is installed on the background thread and long loading dags are not killed

**Comment by ldct** [comment]
To address concerns that timeout won't work on child thread, I've modified the `timeout` class to use `sys.settrace` if it's not on a main thread. I've tested that it times out simple loops and some long-loading airbnb production dags. Disadvantages are that 
1. `settrace`, although part of the public api, is poorly documented
2. it makes loading refreshed dags 2-5x slower (load time on startup is not changed)

**Comment by ldct** [comment]
Closing in favor of https://github.com/apache/incubator-airflow/pull/1685

**Pull request closed without merging**